File: packages/core/src/resources/banking.ts
import { ValidationError } from "../utils/errors";
import { BaseResource } from "./base";
import type {
  BankingComparisonRequest,
  BankingComparisonResponse,
} from "../types";

export class Banking extends BaseResource {
  async compare(
    request: BankingComparisonRequest,
  ): Promise<BankingComparisonResponse> {
    this.validateBankingRequest(request);
    const toolResult = await this.client.callTool(
      "find_banking_accounts",
      request,
    );
    const data =
      this.parseJsonResponse<Omit<BankingComparisonResponse, "widget">>(
        toolResult,
      );
    return {
      ...data,
      widget: this.getWidget(toolResult),
    };
  }

  private validateBankingRequest(request: BankingComparisonRequest): void {
    if (!request.accountType) {
      throw new ValidationError("Account type is required");
    }

    const validTypes = ["checking", "savings", "both"];
    if (!validTypes.includes(request.accountType)) {
      throw new ValidationError(
        `Invalid account type. Must be one of: ${validTypes.join(", ")}`,
      );
    }
  }
}

File: packages/python/securelend/resources/banking.py
from ..utils.errors import ValidationError
from .base import BaseResource
from ..types import BankingComparisonRequest, BankingComparisonResponse

class Banking(BaseResource):
    async def compare(self, request: BankingComparisonRequest) -> BankingComparisonResponse:
        self._validate_banking_request(request)
        tool_result = await self._client.call_tool(
            "find_banking_accounts",
            request,
        )
        data = self._parse_json_response(tool_result)
        data["widget"] = self._get_widget(tool_result)
        return data

    def _validate_banking_request(self, request: BankingComparisonRequest) -> None:
        account_type = request.get("accountType")
        if not account_type:
            raise ValidationError("Account type is required")

        valid_types = ["checking", "savings", "both"]
        if account_type not in valid_types:
            raise ValidationError(
                f"Invalid account type. Must be one of: {', '.join(valid_types)}"
            )
